modules/ethereum-contract/builtin/src/lib.rs

```diff
@@ -45,35 +49,266 @@ pub struct Header {
 }
 
 /// Verify GRANDPA finality proof.
 pub fn verify_substrate_finality_proof(
-	_best_set_id: u64,
-	_raw_best_voters: &[u8],
-	_raw_best_header: &[u8],
-	_raw_headers: &[&[u8]],
-	_raw_finality_proof: &[u8],
-) -> Result<(usize, usize), Error> {
-	Err(Error::JustificationVerify(ClientError::Msg(
-		"Not yet implemented".into(),
-	))) // TODO: implement me
+	finality_target_number: BlockNumber,
+	finality_target_hash: Hash,
+	best_set_id: u64,
+	raw_best_set: &[u8],
+	raw_finality_proof: &[u8],
+) -> Result<(), Error> {
+	let best_set = AuthorityList::decode(&mut &raw_best_set[..]).map_err(Error::BestSetDecode)?;
+	sc_finality_grandpa::GrandpaJustification::<Block>::decode_and_verify_finalizes(
+		&raw_finality_proof,
+		(finality_target_hash, finality_target_number),
+		best_set_id,
+		&best_set.into_iter().collect(),
+	)
+	.map_err(Error::JustificationVerify)
+	.map(|_| ())
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use bridge_node_runtime::DigestItem;
+	use sp_core::crypto::Public;
+	use sp_finality_grandpa::{AuthorityId, ScheduledChange};
+	use sp_runtime::generic::Digest;
+
+	#[test]
+	fn to_substrate_block_number_succeeds() {
+		assert_eq!(to_substrate_block_number(U256::zero()).unwrap(), 0);
+		assert_eq!(
+			to_substrate_block_number(U256::from(std::u32::MAX as u64)).unwrap(),
+			0xFFFFFFFF
+		);
+	}
+
+	#[test]
+	fn to_substrate_block_number_fails() {
+		assert!(matches!(
+			to_substrate_block_number(U256::from(std::u32::MAX as u64 + 1)),
+			Err(Error::BlockNumberDecode)
+		));
+	}
+
+	#[test]
+	fn from_substrate_block_number_succeeds() {
+		assert_eq!(from_substrate_block_number(0).unwrap(), U256::zero());
+		assert_eq!(
+			from_substrate_block_number(std::u32::MAX).unwrap(),
+			U256::from(std::u32::MAX)
+		);
+	}
+
+	#[test]
+	fn substrate_header_without_signal_parsed() {
+		assert_eq!(
+			parse_substrate_header(
+				&RuntimeHeader {
+					parent_hash: [0u8; 32].into(),
+					number: 0,
+					state_root: "b2fc47904df5e355c6ab476d89fbc0733aeddbe302f0b94ba4eea9283f7e89e7"
+						.parse()
+						.unwrap(),
+					extrinsics_root: "03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314"
+						.parse()
+						.unwrap(),
+					digest: Default::default(),
+				}
+				.encode()
+			)
+			.unwrap(),
+			Header {
+				hash: "afbbeb92bf6ff14f60bdef0aa89f043dd403659ae82665238810ace0d761f6d0"
+					.parse()
+					.unwrap(),
+				parent_hash: Default::default(),
+				number: 0,
+				signal: None,
+			},
+		);
+	}
+
+	#[test]
+	fn substrate_header_with_signal_parsed() {
+		let authorities = vec![
+			(AuthorityId::from_slice(&[1; 32]), 101),
+			(AuthorityId::from_slice(&[3; 32]), 103),
+		];
+		let mut digest = Digest::default();
+		digest.push(DigestItem::Consensus(
+			GRANDPA_ENGINE_ID,
+			ConsensusLog::ScheduledChange(ScheduledChange {
+				next_authorities: authorities.clone(),
+				delay: 8,
+			})
+			.encode(),
+		));
+
+		assert_eq!(
+			parse_substrate_header(
+				&RuntimeHeader {
```

Comment:
Can you please add one more comparison against hex to make sure we never break backward compatibility?